src/gateway/bridge/shard_queuer.rs

```diff
@@ -253,3 +291,51 @@ impl ShardQueuer {
         }
     }
 }
+
+/// A queue of [`ShardId`]s that is split up into multiple buckets according to the value of
+/// [`max_concurrency`](crate::model::gateway::SessionStartLimit::max_concurrency).
+#[must_use]
+pub struct ShardQueue {
+    buckets: HashMap<u16, VecDeque<ShardId>>,
+    max_concurrency: NonZeroU16,
+}
+
+impl ShardQueue {
+    pub fn new(max_concurrency: NonZeroU16) -> Self {
+        Self {
+            buckets: HashMap::with_capacity(max_concurrency.get() as usize),
+            max_concurrency,
+        }
+    }
+
+    /// Calculates the corresponding bucket for the given `ShardId` and **appends** to it.
+    pub fn push_back(&mut self, shard_id: ShardId) {
+        let bucket = shard_id.0 % self.max_concurrency.get();
+        self.buckets.entry(bucket).or_default().push_back(shard_id);
+    }
+
+    /// Calculates the corresponding bucket for the given `ShardId` and **prepends** to it.
+    pub fn push_front(&mut self, shard_id: ShardId) {
+        let bucket = shard_id.0 % self.max_concurrency.get();
+        self.buckets.entry(bucket).or_default().push_front(shard_id);
+    }
+
+    /// Pops a `ShardId` from every bucket containing at least one and returns them all as a `Vec`.
+    pub fn pop_batch(&mut self) -> Vec<ShardId> {
+        (0..self.max_concurrency.get())
+            .filter_map(|i| self.buckets.get_mut(&i).and_then(|bucket| bucket.pop_front()))
+            .collect()
```

Comment:
Non blocker: Using a HashMap here seems kind of inefficient, is there a better data structure like BTreeMap that would be automatically sorted you can use?

Reply:
In what sense would a `BTreeMap` be more efficient? Insertions and lookups are constant for HashMaps, not so for `BTreeMap` - and I don't really care about the order of buckets here.

Reply:
Yeah, fair enough, I guess I saw iteration and hashmap and thought inefficient..